src/main/java/com/google/idea/perf/TracerView.kt

```diff
@@ -16,20 +16,86 @@
 
 package com.google.idea.perf
 
+import com.intellij.openapi.application.ApplicationManager.getApplication
 import com.intellij.openapi.ui.MessageType
 import com.intellij.openapi.ui.popup.util.PopupUtil
 import com.intellij.ui.components.JBLabel
 import com.intellij.ui.components.JBPanel
 import com.intellij.util.textCompletion.TextFieldWithCompletion
+import java.awt.event.KeyEvent
+import java.awt.event.KeyListener
 import javax.swing.JProgressBar
 
 /** The content for filling the tracer dialog window. */
 abstract class TracerView: JBPanel<TracerView>() {
+    companion object {
+        const val COMMAND_HISTORY_LIMIT = 20
+    }
+
+    abstract val controller: TracerController
     abstract val commandLine: TextFieldWithCompletion
     abstract val progressBar: JProgressBar
     abstract val refreshTimeLabel: JBLabel
 
+    private val commandHistory = ArrayList<String>(COMMAND_HISTORY_LIMIT)
```

Comment:
I see somewhat buggy behavior when I try’s  using command history, but I'm not sure where the problem lies.

For example, I tried:
```
trace asdf
<up>
<up>
<down>
```
The first `<up>` correctly showed `trace asdf`, then the second `<up>` cleared it, then the `<down>` didn't do anything (I expected it to bring back `trace asdf`).

Maybe I don't know what the intended behavior is, though. I think users will expect the behavior to similar to the "Evaluation Expression" dialog in the debugger, for example.

Reply:
Fixed! I must have made a typo in my code. Recalling the next command checked for `VK_UP` instead of `VK_DOWN`.